internal/pkg/deploy/upload/customresource/customresource.go

```diff
@@ -0,0 +1,185 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package customresource provides functionality to upload Copilot custom resources.
+package customresource
+
+import (
+	"archive/zip"
+	"bytes"
+	"fmt"
+	"io"
+	"path"
+	"strings"
+
+	"github.com/aws/copilot-cli/internal/pkg/template/artifactpath"
+
+	"github.com/aws/copilot-cli/internal/pkg/template"
+)
+
+// Directory under which all custom resources are minified and packaged.
+const customResourcesDir = "custom-resources"
+
+// All custom resource scripts will be copied under this path in the zip file.
+const handlerFileName = "index.js"
+
+// Function names.
+const (
+	envControllerFnName       = "EnvControllerFunction"
+	dynamicDesiredCountFnName = "DynamicDesiredCountFunction"
+	backlogPerTaskFnName      = "BacklogPerTaskCalculatorFunction"
+	rulePriorityFnName        = "RulePriorityFunction"
+	nlbCustomDomainFnName     = "NLBCustomDomainFunction"
+	nlbCertValidatorFnName    = "NLBCertValidatorFunction"
+	customDomainFnName        = "CustomDomainFunction"
+	certValidationFnName      = "CertificateValidationFunction"
+	dnsDelegationFnName       = "DNSDelegationFunction"
+)
+
+// Function source file locations.
+var (
+	albRulePriorityGeneratorFilePath = path.Join(customResourcesDir, "alb-rule-priority-generator.js")
+	backlogPerTaskCalculatorFilePath = path.Join(customResourcesDir, "backlog-per-task-calculator.js")
+	customDomainFilePath             = path.Join(customResourcesDir, "custom-domain.js")
+	customDomainAppRunnerFilePath    = path.Join(customResourcesDir, "custom-domain-app-runner.js")
+	desiredCountDelegationFilePath   = path.Join(customResourcesDir, "desired-count-delegation.js")
+	dnsCertValidationFilePath        = path.Join(customResourcesDir, "dns-cert-validator.js")
+	dnsDelegationFilePath            = path.Join(customResourcesDir, "dns-delegation.js")
+	envControllerFilePath            = path.Join(customResourcesDir, "env-controller.js")
+	nlbCertValidatorFilePath         = path.Join(customResourcesDir, "nlb-cert-validator.js")
+	nlbCustomDomainFilePath          = path.Join(customResourcesDir, "nlb-custom-domain.js")
+)
+
+// CustomResource represents a CloudFormation custom resource backed by a Lambda function.
+type CustomResource struct {
+	name  string
+	files []file
+}
+
+// FunctionName is the name of the Lambda function.
+func (cr CustomResource) FunctionName() string {
```

Comment:
My concern is that the value can get large with the list of files, in which case it might not be the most efficient to use value receiver. Do you think we can use a pointer receiver here?

Reply:
Done for both!